### ISSUE TITLE
Getting relationships logs 

### ISSUE THREAD

**Issue by woliveirac** [spatie/laravel-activitylog#468]
Hello, 

i'm not sure if i can do this already, i've read the docs and all the related issues for this ... but, i couldn't find a solution or at least, not a good one. 

My problem is that i have a model Client and another model ClientStrategy which is a HasMany relationship and i need to get all the changes made to the Client which include the ClientStrategies. 

The thing is, i want to do this: 

Activities::where($client)->get(); // returns all the logs including the relationships changes (insert, delete, update)

ps: // just to give an idea, i know this code isn't the right way of doing this

In my activity_log table ... i have:

![image](https://user-images.githubusercontent.com/7048196/50650599-1bdd1580-0f68-11e9-9dc1-cede56bc52bf.png)

I need to list those  changes exactly like that in the table, but, i can't figured out how can i do that using a where clause that is **"client_id = $x"**, because, i can't query like:

Activities::where('client_id',$id)->get(); // just to give an idea, i know this code isn't the right way of doing this

For now, i'm using the properties and querying like this:

![image](https://user-images.githubusercontent.com/7048196/50650743-78d8cb80-0f68-11e9-9be2-d46e87c711a0.png)

I could think of a few ways to get this going overwriting some things, but, i don't want to get to that point. 

Is there any way of doing this with one query only or at least the least queries possible and avoiding loops? 

Thank you.

**Comment by Gummibeer** [comment]
Ok, first I want to resume what I understood and hope that this is what you need to do!?

You have two models `Client` & `ClientStrategy`. They are related to each other like `Client 1:n ClientStrategy`.
Your goal is to query activities that have one client or related client strategies as subject.

Am I right so far?

If you don't have to do it in one query (even this should be possible) you can do:

_switching from phone to desktop_

**Comment by Gummibeer** [comment]
Hey,

happy to help! 😊
Your two columns could solve your problem, they should be nullable morph columns (take a look in the laravel docs). To fill them you can use an event listener - #462 & #466 as reference.

Related to your last idea: I wouldn't alter the current behavior. I would add a `forSubjectParent()` scope and if you want it in a single scope a second `forSubjectAndParent()`. Every method you override could lead to conflicts in future versions.

All these things would be possible without overriding any package logic. Just extend the model and change it in the config. Add an observer and introduce new methods in the new extending model.

And last: I wouldn't wait for a problem, these changes are simple now. Later you would have to fill the new columns via a migration and so on. So do it better now the "right" way instead of waiting for problems. 😉
Primary if you have 8 childs - logs will explode very fast and in this case my first solution will be slow as hell.

**Comment by woliveirac** [comment]
@Gummibeer Well, that's a good idea, i'll try to get this to work and i let you know the results.

For now, i think you could mark as solved this issue.

Thank you!

**Comment by woliveirac** [comment]
Hey there, it's me again.

I'm trying to do what we talked above.

The thing is ... i can't figure out how can i retrive a "parent" model of any model instance. 

For example, i want to do something like this:

ClientStrategy::findOrFail(15)->getParent(); // returns the client model because ClientStrategy belongsTo Client model. 

i could write a method which would give me that information, yes, but, the thing is that i need that in a generic form because i'm using it on my observer ... that is listening to the **Activity::updating** event. 

The whole scenario right now is: 

I have created the **ActivityCustom** model which **Extends** **Activity** 

![image](https://user-images.githubusercontent.com/7048196/51405102-897a6b80-1b3c-11e9-8c6a-f3c20159131c.png)

I have added the nullableMorphs fields that you have mentioned and now i have this table structure:
![image](https://user-images.githubusercontent.com/7048196/51405163-bc246400-1b3c-11e9-828b-1d487a053cfa.png)

So now, in my Observer, i need to fill the nullableMorphs fields with the parent_id and parent_type ... but, i have no idea how can i do that ...

![image](https://user-images.githubusercontent.com/7048196/51405224-e70eb800-1b3c-11e9-8ad8-25ea95c8c69e.png)

Client model has: 

![image](https://user-images.githubusercontent.com/7048196/51405272-0c9bc180-1b3d-11e9-978d-d2da884d3b8e.png)

and the ClientStrategy model is:

![image](https://user-images.githubusercontent.com/7048196/51405314-289f6300-1b3d-11e9-8e4c-0882cfc1ae70.png)

How can i fill those two fields generically inside the Observer ... ? 

I thought about doing something like this, but, doesn't work: 

![image](https://user-images.githubusercontent.com/7048196/51405523-acf1e600-1b3d-11e9-9646-6c1c7c674766.png)

Any ideas? What am i doing wrong?

Thank you!

**Comment by Gummibeer** [comment]
Hey,

First I would like to point you to #472 - this will allow you to set these fields from the `ClientStrategy::tapActivity()` and not an external observer.

This will also prevent you from creating a dynamic `getParent()` but you can use the existing relationship method.

Now you have to add the `parent` relation to your custom activity model. This will allow you to do following:

*ClientStrategy*
```php
$activity->parent()->associate($this->client);
```

Otherwise I would add a switch case in the observer:

```php
switch(get_class($subject)) {
  case ClientStrategy::class:
    $activity->parent()->associate($subject->strategy);
    break;
}
```

You should prevent setting these attributes your own because there is a MorphMap in laravel.

**Comment by woliveirac** [comment]
Hello,

well, thank you very much @Gummibeer , got this working. 

I got rid of the Observer and instead i used the TapActivity that you have pointed #472 .

I will post the steps just for the people curious about  this issue. 

Also, thank you pointing out the MorphMap, it'll be really useful in the future. 

PS: I had to pull the "ft-activity-tap" branch because it isn't in the Master yet (this got me thinking for a second why isn't it working haha) 

So, it's really easy to get this working:

![image](https://user-images.githubusercontent.com/7048196/51476469-c9279a00-1d6c-11e9-8f53-25c382f65dce.png)

![image](https://user-images.githubusercontent.com/7048196/51476485-d17fd500-1d6c-11e9-8e6d-2b5a9446fee9.png)

Result: 

![image](https://user-images.githubusercontent.com/7048196/51476505-e0ff1e00-1d6c-11e9-9d3e-899a4585729a.png)

Thank you again.

**Comment by Gummibeer** [comment]
@woliveirac yes, atm you have to use `dev-ft-activity-tap` as composer version.
Thanks for your feedback! :) 
Can you please add an approve review to the PR #472 ? This will help me to keep track of the approves in one place.

And just a personal hint, possible too late, I would rename `parent` to something like `main_subject`. Parent is a reserved keyword in PHP which could lead to problems in some circumstances. And it's not this absurd that Laravel could introduce a method/property with this name, or any other package.

**Comment by woliveirac** [comment]
Well, thanks for the tip @Gummibeer , already changed its name.

About the review in the PR, is it just a comment? or there is something or a process to do? 

Thanks!

**Comment by olivermbs** [comment]
Thank you for this useful thread! Just one question, I get the following error (my code is otherwise the same):

`App\Models\TripQuote::tapActivity(): Argument #1 ($activityCustom) must be of type App\Models\ActivityCustom, Spatie\Activitylog\Models\Activity given`

Any ideas?

**Comment by kindlebug** [comment]
@olivermbs hello i got similiar issue on this ($activityCustom)

App\Models\TripQuote::tapActivity(): Argument #1 ($activityCustom) must be of type App\Models\ActivityCustom, Spatie\Activitylog\Models\Activity given

any screen shot to how publish the the config file and override the model ?

**Comment by olivermbs** [comment]
> @olivermbs hello i got similiar issue on this ($activityCustom)
> 
> App\Models\TripQuote::tapActivity(): Argument #1 ($activityCustom) must be of type App\Models\ActivityCustom, Spatie\Activitylog\Models\Activity given
> 
> any screen shot to how publish the the config file and override the model ?

Publish the config file with:

`php artisan vendor:publish --provider="Spatie\Activitylog\ActivitylogServiceProvider" --tag="activitylog-config"`

Then in this file set the activity_model:

` 'activity_model' => \App\Models\ActivityCustom::class,`
